### DISCUSSION TITLE
Binding API that uses direct references to objects

### DISCUSSION THREAD

**Discussion by jeffective** [zauberzeug/nicegui#3080] [Ideas / Feature Requests]
Currently, the binding API requires providing an object and a string. One footgun with this approach is that it is very easy to have a typo in the string.

Could we perhaps have an API that takes a direct reference to the object to bind to?

The existing api:

```python
@dataclass
class BindableData:
    foo: str = "Bar"


my_bindable_data = BindableData()

ui.input().bind_value_from(my_bindable_data, "foo")
ui.input().bind_value_from(my_bindable_data, "foo1") # oops! typo!

```
The api could be significantly easier to use if the following was possible:

```python
@dataclass
class BindableData:
    foo: str = "Bar"


my_bindable_data = BindableData()

ui.input().bind_value_from(my_bindable_data.foo) # my type checker / autocomplete prevented the typo!
```

Of course the developer would need to be careful to preserve this reference (which we already have to do since there are already existing footguns related to overwriting these references to existing objects and breaking the binding connection).

**Comment by BlankAdventure** [comment]
I don't have much to add other than to say I have had the same thought.

**Comment by falkoschindler** [comment]
Great idea, @kj4tmp!

That's exactly what we had in mind when designing NiceGUI's binding API around three years ago. We thought it would feel much _nicer_ than fiddling with strings. And we actually implemented it as a separate library https://github.com/zauberzeug/binding/ and integrated it in NiceGUI 0.2.0: 3915a3d2cecc2c136faf92d4b7559ff3d52c61f4.

But as described in https://github.com/zauberzeug/binding?tab=readme-ov-file#implementation-and-dependencies, there were several tricks required to get this working. Especially relying on the [inspect](https://docs.python.org/3/library/inspect.html) and [executing](https://pypi.org/project/executing/) packages to detect object and attribute name felt a little unstable. When a binding function receives a "source" or "target", it usually only sees the value (e.g. `str` or `int`) but not the corresponding object and the attribute name. Therefore it needed to inspect the callee and look for this information itself:
```py
def _get_parent_and_argument(method_name, nesting):

    sep = f'.{method_name}('
    expression = inspect.stack()[2].code_context[0].strip().split(sep)[0]
    callFrame = inspect.currentframe().f_back.f_back
    self_obj, self_name = _extract_object(expression, callFrame)

    for _ in range(nesting):
        callFrame = callFrame.f_back
    callNode = Source.executing(callFrame).node
    source = Source.for_frame(callFrame)
    expression = source.asttokens().get_text(callNode.args[0])
    other_obj, other_name = _extract_object(expression, callFrame)

    return self_obj, self_name, other_obj, other_name
```
https://github.com/zauberzeug/binding/blob/8b09b15e3675c68850b81d7253f1eb24245ee76b/binding/binding.py#L15-L29

This allows to write:
```py
label.text.bind_from(input.value)
```
But it breaks as soon as the user deviates from the intended usage, e.g. when writing:
```py
first_name = input.value
label.text.bind_from(first_name)
```

And another problem with this fancy API was that `label.text` is a `str` object which has no `bind_from` method. So we used the [forbiddenfruit](https://pypi.org/project/forbiddenfruit/) package to extend ("curse") built-in types like `str` and `int` with our binding methods. It worked, but didn't feel quite right for a stable production-ready UI library. That's why we abandoned this library soon after.

Long story short, this API felt very compelling, but required nasty hacks to ultimately work.

**Reply by jeffective** [reply]
Thank you for taking the time to explain! I expected there were language feature limitations / python intricacies that encouraged the string binding API.
